src/transformation/transformers/access.ts

```diff
@@ -0,0 +1,89 @@
+import * as ts from "typescript";
+import * as tstl from "../../LuaAST";
+import { transformBuiltinPropertyAccessExpression } from "../builtins";
+import { FunctionVisitor, TransformationContext, TransformerPlugin } from "../context";
+import { DecoratorKind, getCustomDecorators } from "../utils/decorators";
+import { createExpressionPlusOne } from "../utils/lua-ast";
+import { isArrayType, isNumberType, isStringType } from "../utils/typescript";
+
+export function transformElementAccessArgument(
+    context: TransformationContext,
+    expression: ts.ElementAccessExpression
+): tstl.Expression {
+    const index = context.transformExpression(expression.argumentExpression);
+
+    const type = context.checker.getTypeAtLocation(expression.expression);
+    const argumentType = context.checker.getTypeAtLocation(expression.argumentExpression);
+    if (isNumberType(context, argumentType) && isArrayType(context, type)) {
+        return createExpressionPlusOne(index);
+    }
+
+    return index;
+}
+
+const transformElementAccessExpression: FunctionVisitor<ts.ElementAccessExpression> = (expression, context) => {
```

Comment:
Unlike declarations, function expressions can have a value type annotation.
Using a value level annotation is shorted (declaration would look like `function transformElementAccessExpression(expression: ts.ElementAccessExpression, context: TransformationContext): tstl.Expression`) and reveals possible errors at a declaration time (instead of at usage for declaration).